### DISCUSSION TITLE
Object tracking without object type identification?

### DISCUSSION THREAD

**Discussion by jtut1731** [blakeblackshear/frigate#3714] [General]
I have been using Frigate for quite awhile, and it beats everything out there. I searched before asking this question, but I can't find any discussion about this directly. Please point me to anything, if it has been discussed. It would also be very helpful in the docs, if "How it works" could be written to explain the process Frigate uses (if I didn't miss it somehow).

I seem to recall from some documentation, that Frigate uses motion detection (using what?), and then those motion boxes are sent to the object detection. If, and only if, there is at least one object detected in the bounding box, will an "object" be considered for further processing (like minimum size, etc.)

The problem with this is becoming apparent to me with the introduction of Frigate+. I have started evaluating the upcoming version 0.11.0. There is a way from the GUI to send the event image to Frigate+ to improve the model. The first thing about this is: frigate could already detect this object if we got this far. The second is what seems to be a lack of "unknown object" detection. For instance, today with 0.10.0, there is not a model for "deer". The only way that I could upload my image of a deer that was captured with frigate, would be to find it in the recording, and send a snapshot of it to frigate+. There is not an easy way that I have seen to upload the image of the deer, because frigate did not create an event, because the deer was not detected as an object.

Would there be a way after detecting motion, and maybe detecting that the bounding box moves, or some other clever detection method, for frigate to create an event of an "unknown object"? This would make it very easy to send the snapshot to Frigate+ for further model training. I may or may not want to alert on "unknown object", but I may never know that there was a rainbow colored unicorn in my front yard, because frigate never captured it as an event.

I would love to know your thoughts.

**Comment by NickM-27** [comment]
> that Frigate uses motion detection (using what?)

Frigate uses `cv2` on the Y channel of the YUV420 encoded images to look for motion. Motion being a change in pixels outside of certain parameters that are set in the config.

> and then those motion boxes are sent to the object detection. If, and only if, there is at least one object detected in the bounding box, will an "object" be considered for further processing (like minimum size, etc.)

The motion boxes in that frame are considered and a "region" is created of those motion boxes. This region is then resized to the models input size (in the default models case it is 320x320 pixels). Sometimes based on the region and the objects position in the region, multiple detections may be run on the same frame.

> The first thing about this is: frigate could already detect this object if we got this far

The first point of Frigate+ is just making the current labels better. The default model is a google test model and the images used to train it were not from a security camera point of view and are also heavily skewed towards `person`. Frigate+ will be trained to make existing detections much more accurate (for example my dog is often detected as cat or person depending on his current walking angle relative to the camera)

> Would there be a way after detecting motion, and maybe detecting that the bounding box moves, or some other clever detection method, for frigate to create an event of an "unknown object"? 

It doesn't really work like that, the model does not know what an object is. It only knows what a "person" looks like, "dog" looks like, etc. so if it doesn't detect those specific ones then it thinks it is nothing. It would be a major rewrite to use motion to try and infer this and there would be loads of false positives anyway making it frustrating to use.

As far as the deer scenario goes, in my experience it is detected as `dog` or `cat` most the time anyway and can be sent to Frigate+ that way.

If you REALLY want to, you could use https://docs.frigate.video/configuration/advanced#model to set unused labels as `unknown` or something but then you'd still just end up with loads of false positives.

**Comment by blakeblackshear** [comment]
The best detailed explanation of how frigate works is here: https://docs.frigate.video/configuration/masks#further-clarification

The best way to identify "unknown objects" is to review motion footage where objects weren't detected. This is something planned for the future since all the data you need is in the database. Reviewing footage of motion, grabbing a single frame, and labeling it for frigate+ will be added in the future.

**Comment by lindhe** [comment]
I too have been looking for a way to detect "any motion", for automation in Home Assistant!

**Reply by NickM-27** [reply]
This is already supported in 0.11

**Reply by lindhe** [reply]
Oh, how neat! I didn't find that in the documentation, do you have a link?

**Reply by NickM-27** [reply]
https://docs.frigate.video/integrations/mqtt#frigatecamera_namemotion

There's a sensor in the home assistant integration as well

**Comment by jtut1731** [comment]
Correct me if I’m wrong, but frigate 0.11 supports “any motion” as motion
with one or more recognized objects. Not really the definition of “any
motion” in my mind.

**Reply by NickM-27** [reply]
You are wrong, Frigate 0.11 supports an MQTT topic / sensor which exposes detected motion to the user. Detected motion meaning a change in pixels. After all, frigate only runs object detection when it has detected motion. 

![SmartSelect_20221003_071914_Home Assistant.jpg](https://user-images.githubusercontent.com/14866235/193587886-84286885-1a87-4447-9f22-26671fd06c5e.jpg)

**Reply by NickM-27** [reply]
It was added in https://github.com/blakeblackshear/frigate/pull/3152

**Comment by jtut1731** [comment]
Awesome! I’ll have to try that myself.

**Comment by sir-smythe** [comment]
If you set an unused object to "unknown" and set the threshold and min_score to 0 would that surface all the detection boxes as an event?
